File: src/RqCalc.Wpf/Models/TalentBranchModel.cs
using Framework.Core;
using Framework.Reactive;

using RqCalc.Domain;
using RqCalc.Domain.Talent;
using RqCalc.Wpf.Models.Window.Dialog;

namespace RqCalc.Wpf.Models;

public class TalentBranchModel : NotifyModelBase
{
    public readonly TalentsWindowModel WindowModel;
    
    public TalentBranchModel(IReadOnlyDictionary<TextTemplateVariableType, decimal> evaluateStats, TalentsWindowModel windowModel, ITalentBranch talentBranch)
    {
        this.WindowModel = windowModel;
        this.TalentBranch = talentBranch;


        var talentRequest = from tal in talentBranch.Talents

            group tal by tal.HIndex into hGroup

            orderby hGroup.Key

            select hGroup;

        this.TalentMatrix = talentRequest.Reverse()
                                         .Select((g, index) => g.OrderBy(tal => tal.VIndex).ToObservableCollection(tal => new TalentModel(evaluateStats, this, index == 0, tal)))
                                         .Reverse()
                                         .ToObservableCollection();

        this.TalentList = this.TalentMatrix.SelectMany().ToObservableCollection();
    }


    public IEnumerable<ITalent> ActiveTalents
    {
        get =>
            from collection in this.TalentMatrix

            from model in collection

            where model.Active

            select model.SelectedObject;
        set
        {
            var cache = value.Pipe(Enumerable.ToHashSet);

            foreach (var collection in this.TalentMatrix)
            {
                foreach (var model in collection)
                {
                    model.Active = cache.Contains(model.SelectedObject);
                }
            }
        }
    }

    public ObservableCollection<TalentModel> TalentList
    {
        get => this.GetValue(v => v.TalentList);
        private set => this.SetValue(v => v.TalentList, value);
    }

    public ObservableCollection<ObservableCollection<TalentModel>> TalentMatrix
    {
        get => this.GetValue(v => v.TalentMatrix);
        private set => this.SetValue(v => v.TalentMatrix, value);
    }

    public ITalentBranch TalentBranch
    {
        get => this.GetValue(v => v.TalentBranch);
        private set => this.SetValue(v => v.TalentBranch, value);
    }
}

File: src/RqCalc.WebApp/_Scripts/Models/TalentBranchModel.js
function TalentBranchModel(context, evaluateStats, windowModel, branch) {

    var thisModel = this;

    thisModel.branch = branch;
    thisModel.window = windowModel;
    //-----------------------------------------------------------------------------------------------------------------
    thisModel.talentMatrix = branch.talents.groupBy(function(tal) {
            return tal.hIndex;
        })
        .orderBy(function(g) {
            return g.key;
        })
        .reverseC()
        .map(function(g, index) {

            return g.value.orderBy(function(tal) {
                    return tal.vIndex;
                })
                .map(function(tal) {
                    return new TalentModel(context, evaluateStats, thisModel, tal, index === 0);
                });
        })
        .reverseC();

    //-----------------------------------------------------------------------------------------------------------------
    thisModel.getTalentModels = function () {

        return thisModel.talentMatrix.selectMany(function(talentColumn) {
            return talentColumn;
        });
    }

    //-----------------------------------------------------------------------------------------------------------------
    thisModel.setActiveTalents = function(talents) {

        thisModel.getTalentModels()
            .foreach(function(talentModel) {
                talentModel.active(talents.contains(talentModel.talent));
            });
    };
    //-----------------------------------------------------------------------------------------------------------------
    thisModel.getActiveTalents = function() {

        return thisModel.getTalentModels()
            .filter(function(talentModel) {
                return talentModel.active();
            })
            .map(function(talentModel) {
                return talentModel.talent;
            });
    };
}